config/chassis_modules.py

```diff
@@ -17,6 +21,74 @@ def modules():
     """Configure chassis modules"""
     pass
 
+
+def get_config_module_state(db, chassis_module_name):
+    config_db = db.cfgdb
+    fvs = config_db.get_entry('CHASSIS_MODULE', chassis_module_name)
+    if not fvs:
+        return 'up'
+    else:
+        return fvs['admin_status']
+
+
+#
+# Name: check_config_module_state_with_timeout
+# return: True: timeout, False: not timeout
+#
+def check_config_module_state_with_timeout(ctx, db, chassis_module_name, state):
+    counter = 0
+    while get_config_module_state(db, chassis_module_name) != state:
+        time.sleep(1)
+        counter += 1
+        if counter >= TIMEOUT_SECS:
+            ctx.fail("get_config_module_state {} timeout".format(chassis_module_name))
+            return True
+    return False
+
+
+def get_asic_list_from_db(chassisdb, chassis_module_name):
+    asic_list = []
+    asics_keys_list = chassisdb.keys("CHASSIS_STATE_DB", "CHASSIS_FABRIC_ASIC_TABLE*")
+    for asic_key in asics_keys_list:
+        name = chassisdb.get("CHASSIS_STATE_DB", asic_key, "name")
+        if name == chassis_module_name:
+            asic_id = int(re.search(r"(\d+)$", asic_key).group())
+            asic_list.append(asic_id)
+    return asic_list
+
+
+#
+# Syntax: fabric_module_set_admin_status <chassis_module_name> <'up'/'down'>
+#
+def fabric_module_set_admin_status(db, chassis_module_name, state):
+    chassisdb = db.db
+    chassisdb.connect("CHASSIS_STATE_DB")
+    asic_list = get_asic_list_from_db(chassisdb, chassis_module_name)
+
+    if len(asic_list) == 0:
+        return
+
+    if state == "down":
+        for asic in asic_list:
+            click.echo("Stop swss@{} and syncd@{} services".format(asic, asic))
+            clicommon.run_command('sudo systemctl stop swss@{}.service'.format(asic))
+
+        # wait for service is down
+        time.sleep(2)
+        click.echo("Delete related CAHSSIS_FABRIC_ASIC_TABLE entries")
+
+        for asic in asic_list:
+            chassisdb.delete("CHASSIS_STATE_DB", "CHASSIS_FABRIC_ASIC_TABLE|asic" + str(asic))
+
+        for asic in asic_list:
+            click.echo("Start swss@{} and syncd@{} services".format(asic, asic))
+            clicommon.run_command('sudo systemctl reset-failed swss@{}.service'.format(asic))
+            clicommon.run_command('sudo systemctl start swss@{}.service'.format(asic))
+    else:
+        for asic in asic_list:
+            click.echo("Start swss@{} and syncd@{} services".format(asic, asic))
```

Comment:
won't we require a platform hook to let platform power on the FC module?

Reply:
By design, it should not be such case -- unless users just execute issue command "systemctl stop  swss@/syncd@" without bring down the hardware.  This handling just covers this case.  
As normal operation,  if user use CLI "config chassis module shutdown " a Fabric module. We shutdown service, delete the CHASSIS_FABRIC_ASIC_TABLE entry, we restart the service since we have the asic_list at this point.  This is the design which we discussed in the meeting that no need the Vendor provide the hook to do it.

Reply:
Are you suggesting that the Vendor does not need to power down the FC? So the slot is always powered up. Once a new FC or old FC is inserted again, it will be powered up but admin down? Won't the asic detected already and inserted in db by pmon module?

Reply:
> Are you suggesting that the Vendor does not need to power down the FC? So the slot is always powered up. Once a new FC or old FC is inserted again, it will be powered up but admin down? Won't the asic detected already and inserted in db by pmon module?

No, FC is power down by the  set_admin_state in module.py (Vendor specified code).  This is an exists functionality which has been done in the chassisd->ModuleConfigUpdater().  If Vendors support shutdown/startup FC, they need to modify their sonic-platform module.py -> set_admin_state(0 function to support it